### PR TITLE
ICU-13850 Make CurrencyUnit safe with 1-length and 2-length strings.

### PR DESCRIPTION
**Pull request by sffc**
Replying to Markus's long-overdue feedback on ICU-13850.

##### Checklist

- [x] Issue filed at https://unicode-org.atlassian.net :  ICU-13850
- [x] Update PR title to include Issue number
- [x] Issue accepted
- [x] Tests included
- [ ] Documentation is changed or added

### CODE REVIEW

**Comment by markusicu** [comment]
Why is intltest TestIcuTestErrorCode failing?

File: icu4c/source/i18n/currunit.cpp [hunk 1]
```diff
@@ -28,9 +28,13 @@ CurrencyUnit::CurrencyUnit(ConstChar16Ptr _isoCode, UErrorCode& ec) {
     // Note: in ICU4J Currency.getInstance(), we check string length for 3, but in ICU4C we allow a
     // non-NUL-terminated string to be passed as an argument, so it is not possible to check length.
     // However, we allow a NUL-terminated empty string, which should have the same behavior as nullptr.
+    // Consider NUL-terminated strings of length 1 or 2 as invalid.
     const char16_t* isoCodeToUse;
     if (U_FAILURE(ec) || _isoCode == nullptr || _isoCode[0] == 0) {
         isoCodeToUse = kDefaultCurrency;
+    } else if (_isoCode[1] == 0 || _isoCode[2] == 0) {
+        isoCodeToUse = _isoCode;
```
Comment (markusicu) [inline icu4c/source/i18n/currunit.cpp]: It looks like error handling sets isoCodeToUse = kDefaultCurrency; so we should do so here, too (as I suggested in my ticket comment).

Optional: With that, other than the TODO comment, you could merge this with the not-invariant check below.

**Comment by sffc** [comment]
I just committed a fix for the failing test.

File: icu4c/source/tools/ctestfw/tstdtmod.cpp [hunk 2]
```diff
@@ -88,12 +111,14 @@ void IcuTestErrorCode::setScope(const UnicodeString& message) {
 }
 
 void IcuTestErrorCode::handleFailure() const {
-    errlog(FALSE, nullptr);
+    errlog(FALSE, u"", nullptr);
 }
 
-void IcuTestErrorCode::errlog(UBool dataErr, const char* extraMessage) const {
+void IcuTestErrorCode::errlog(UBool dataErr, UnicodeString mainMessage, const char* extraMessage) const {
```
Comment (markusicu) [inline icu4c/source/tools/ctestfw/tstdtmod.cpp]: Please use a const reference for mainMessage.

File: icu4c/source/tools/ctestfw/tstdtmod.cpp [hunk 3]
```diff
@@ -88,12 +111,14 @@ void IcuTestErrorCode::setScope(const UnicodeString& message) {
...
     UnicodeString msg(testName, -1, US_INV);
-    msg.append(u" failure: ").append(UnicodeString(errorName(), -1, US_INV));
+    msg.append(u" ");
```
Comment (markusicu) [inline icu4c/source/tools/ctestfw/tstdtmod.cpp]: Optional: You could use the UChar append and chain the string append:

msg.append(u' ').append(mainMessage);

File: icu4c/source/i18n/currunit.cpp (hunk 1 above)
Comment (sffc) [reply icu4c/source/i18n/currunit.cpp]: > It looks like error handling sets isoCodeToUse = kDefaultCurrency; so we should do so here, too (as I suggested in my ticket comment).

Done; good catch.

> Optional: With that, other than the TODO comment, you could merge this with the not-invariant check below.

I made the error codes different.  This one is U_ILLEGAL_ARGUMENT_ERROR, and the other case uses U_INVARIANT_CONVERSION_ERROR.  So it's cleaner to have them be separate cases of the if statement.

File: icu4c/source/tools/ctestfw/tstdtmod.cpp (hunk 2 above)
Comment (sffc) [reply icu4c/source/tools/ctestfw/tstdtmod.cpp]: Done.

File: icu4c/source/tools/ctestfw/tstdtmod.cpp (hunk 3 above)
Comment (sffc) [reply icu4c/source/tools/ctestfw/tstdtmod.cpp]: Done.

**Comment by sffc** [comment]
Sorry, I wanted to add more test cases.  Please see the latest commit and re-approve.

**Pull request merged**